### PR TITLE
Symbolic variance and covariance functions

### PR DESCRIPTION
**Pull request by Upabjojr**
instead of evaluating the integral to calculate variance and covariance, keep them in _hold mode_, similar to the relationship between _Integral_ and _integrate_.

### CODE REVIEW

File: sympy/stats/rv_interface.py [hunk 1]
```diff
@@ -79,6 +125,69 @@ def standard_deviation(X, condition=None, **kwargs):
...
+        rv = []
+        nonrv = []
+        for a in m.args:
+            if a.has(RandomSymbol):
+                rv.append(a)
+            else:
+                nonrv.append(a)
+        return (Mul(*nonrv), Mul(*rv))
+
+    def doit(self, **kwargs):
```
Comment (scopatz) [inline sympy/stats/rv_interface.py]: Needs a docstring.

**Comment by scopatz** [comment]
Thanks for submitting this @Upabjojr

File: sympy/stats/rv_interface.py (hunk 1 above)
Comment (Upabjojr) [inline sympy/stats/rv_interface.py]: Yes, I know. Also doctests. I wanted some community feedback first.

**Comment by Upabjojr** [comment]
Hi @mrocklin , I added these two expression to handle symbolic variance and covariance expression. If you have time, tell me what you think.

**Comment by arita291** [comment]
Hello,

I think this is a good idea to keep higher level of abstraction, it gives flexibility for algebric manipulation.
Thank you
Kevin

**Comment by Upabjojr** [comment]
@leosartaj Since you mentioned this PR, what do you think about it?

**Comment by leosartaj** [comment]
> @leosartaj Since you mentioned this PR, what do you think about it?

I feel this is a good start and is certainly something that will be useful. I am not very familiar with the stats code, but I will try to look through it.

**Comment by Upabjojr** [comment]
@leosartaj notice it is being pulled into the sandbox module, which is specifically intended for work-in-progress code. I feel like this code could be used in the other PR.

**Comment by leosartaj** [comment]
> @leosartaj notice it is being pulled into the sandbox module, which is specifically intended for work-in-progress code.

I noticed.

> I feel like this code could be used in the other PR.

+1, I think the code should be tried and made to port to the stats package. I do not see a reason why this is not a good idea.

File: sympy/sandbox/symbolic_probability.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+
+class Probability(Expr):
+    """
+    Symbolic expression for the probability.
+    """
+    def __new__(cls, prob, condition=None, **kwargs):
+        prob = _sympify(prob)
+        if condition is None:
+            obj = Expr.__new__(cls, prob, **kwargs)
```
Comment (jksuom) [inline sympy/sandbox/symbolic_probability.py]: It seems that ``Expr.__new__`` does not accept keyword arguments.

**Comment by Upabjojr** [comment]
> code should be tried and made to port to the stats package.

I do not feel like merging this in the _stats_ package, we don't have a way to express independence/correlation between random variables, and it's likely to take a long time to introduce.

I suggest to merge this into the sandbox, so we can experiment without worrying about API breaks.

**Comment by leosartaj** [comment]
> I do not feel like merging this in the stats package, we don't have a way to express independence/correlation between random variables, and it's likely to take a long time to introduce.

> I suggest to merge this into the sandbox, so we can experiment without worrying about API breaks.

+1

**Comment by Upabjojr** [comment]
Spurious failure, can someone restart Travis?

**Comment by asmeurer** [comment]
Why is this in sandbox?

**Comment by Upabjojr** [comment]
> Why is this in sandbox? 

So we don't have to worry about future API break when we will try to merge this functionality with the existing one in _stats_.

Furthermore we need to find a way to represent dependence/independence, _sandbox_ is good to experiment without fearing API breaks.

**Comment by Upabjojr** [comment]
@asmeurer do you feel like merging this straight into the _stats_ module?

**Comment by leosartaj** [comment]
> So we don't have to worry about future API break when we will try to merge this functionality with the existing one in stats.

> Furthermore we need to find a way to represent dependence/independence, sandbox is good to experiment without fearing API breaks.

I agree with you on this. But I would like to see this eventually getting merged into the stats package. Probably when everybody is happy with the new additions. I feel this PR does bring in desired functionality.

**Comment by asmeurer** [comment]
If you still feel unclear that this is the best API then we can leave it in sandbox. It should eventually go in stats but it's fine to let the API clear up first.

**Comment by asmeurer** [comment]
Although that raises the question for https://github.com/sympy/sympy/pull/10960/. I feel like code outside of sandbox shouldn't depend on code inside of sandbox in a user-facing way (i.e., the code outside would have to API break if the sandbox code API breaks). Depending on it only internally is fine because internal-only API breaks are not a big deal. If I understand your suggestions correctly, you want to use this in https://github.com/sympy/sympy/pull/10960/ in a user-facing way (return functions from this module from variance_prop).

**Comment by Upabjojr** [comment]
@asmeurer you have convinced me. Moved file back to the _stats_ module.

**Comment by asmeurer** [comment]
I think what's here looks fine. If you don't plan on adding anything else, you can go ahead and merge it. 

But as I was saying on #10960, it would be useful to be able to define a normal random variable without specifying a mean or standard deviation, like `RandomSymbol('X')` (not sure if we should reuse the name `RandomSymbol`) and have those simply be represented as unevaluated `Expectation(X)` and `sqrt(Variance(X))`. If I understand the [math](https://en.wikipedia.org/wiki/Propagation_of_uncertainty) behind behind variance propogation, there is no assumption on the type of the random variable (i.e., it doesn't necessarily have to be normal). 

For #10960 there's also the question of whether we want to force users to use RandomSymbols for random variables. It seems there should be some way to just use normal symbols and do the right thing. 

I think this can be merged now and we can figure these things out in another pull request, though.

**Comment by scopatz** [comment]
I agree with @asmeurer.  Not necessarily to be done here, but I think a natural extension to this work would be to add skew, kurtosis, and perhaps a kth order moment object, which it would be nice if some of the current Expectation and Variance classes were subclasses of.  I think that is beyond the scope here, but just wanted to throw it out there. https://en.wikipedia.org/wiki/Standardized_moment

**Comment by Upabjojr** [comment]
> I think a natural extension to this work would be to add skew, kurtosis, and perhaps a kth order moment object,

Agree, but in another PR.

> which it would be nice if some of the current Expectation and Variance classes were subclasses of.

Why subclasses? Skewness and kurtosis, although related, are difference concepts than expectation and variance.

**Comment by Upabjojr** [comment]
This PR misses the algebraic properties of the Expectation, but better add that in another PR. I'm +1 to merge this.

**Comment by Upabjojr** [comment]
Let's merge this next Monday, if nobody opposes it.

File: sympy/stats/symbolic_probability.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+            outval = []
+            for a in expr.args:
+                rv = []
+                nonrv = []
+                if isinstance(a, Mul):
+                    outval.append(cls._get_mul_nonrv_rv_tuple(a))
+                elif isinstance(a, RandomSymbol):
+                    outval.append((S.One, a))
+
+                outval.append((Mul(*nonrv), Mul(*rv)))
```
Comment (jksuom) [inline sympy/stats/symbolic_probability.py]: Under which conditions would ``nonrv`` or ``rv`` be non-empty? Is something missing?

Comment (Upabjojr) [inline sympy/stats/symbolic_probability.py]: Yeah, this line is useless.

Comment (Upabjojr) [inline sympy/stats/symbolic_probability.py]: `nonrv` and `rv` where later moved to `_get_mul_nonrv_rv_tuple`, I forgot to delete them here.

**Comment by jksuom** [comment]
It is possible to define
```
In [6]: X = Normal('X', 2, 3)
In [7]: var = Variance(X**2, evaluate=False)
In [8]: var.doit()
Out[8]: 306
```
but ``Variance(X**2)`` raises a ``NotImplementedError``. Could it be possible to just return the unevaluated ``Variance`` also here?

**Comment by Upabjojr** [comment]
OK, let's merge this at the beginning of may if no further issues arise.

**Comment by jksuom** [comment]
I think this looks good. I'll merge in 24 hrs if there are no more comments.

**Pull request merged**